src/offset_set.rs

```diff
@@ -0,0 +1,193 @@
+        // - using `leb128` encoding might not make sense at all for types with larger alignment
+        // - otherwise this might be missing a couple of places that need explicit alignment
+        // - and as we have found out, miri is particularly picky about alignment as well :-)
+        assert!(
+            mem::align_of::<T>() == 1,
+            "T is currently limited to alignment `1`"
+        )
+    };
+
+    /// Initializes an empty [`OffsetSet`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns the slice stored at the given offset in the byte slice, if any.
+    ///
+    /// Use this to retrieve a slice that was previously [inserted](OffsetSet::insert) into an [`OffsetSet`].
+    pub fn read(buffer: &[u8], offset: usize) -> Result<&[T], ReadOffsetSetError> {
+        Ok(OffsetSet::read_internal(buffer, offset)?.0)
+    }
+
+    fn read_internal(buffer: &[u8], offset: usize) -> Result<(&[T], usize), ReadOffsetSetError> {
+        let mut cursor = Cursor::new(
+            buffer
+                .get(offset..)
+                .ok_or(ReadOffsetSetError::OutOfBounds)?,
+        );
+        let len = leb128::read::unsigned(&mut cursor)? as usize;
+        // it would be nice if `leb128` would directly return this as well,
+        // so one wouldn't have to use a `Cursor`.
+        let leb_len = cursor.position() as usize;
+
+        let start = offset + leb_len;
+        let end = start + len * mem::size_of::<T>();
+
+        let bytes = buffer
+            .get(start..end)
+            .ok_or(ReadOffsetSetError::OutOfBounds)?;
+        let slice = T::slice_from_bytes(bytes).ok_or(ReadOffsetSetError::OutOfBounds)?;
+
+        Ok((slice, end))
+    }
+
+    /// Iterates over all the entries is this [`OffsetSet`].
+    ///
+    /// This yields `(offset, slice)` pairs.
+    pub fn entries(&self) -> impl Iterator<Item = (usize, &[T])> + '_ {
+        self.offsets
+            .iter()
+            .map(|&offset| (offset, Self::read(&self.buffer, offset).unwrap()))
+    }
+
+    /// Returns a byte slice containing the serialized representation of this [`OffsetSet`].
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.buffer
+    }
+
+    /// Returns a byte vector containing the serialized representation of this [`OffsetSet`].
+    ///
+    /// This consumes the [`OffsetSet`].
+    pub fn into_bytes(self) -> Vec<u8> {
+        self.buffer
+    }
+}
+
+impl<T: Pod + PartialEq + Hash> OffsetSet<T> {
+    fn raw_entry(&mut self, items: &[T]) -> (Entry<'_, usize>, &mut Vec<u8>) {
+        let hasher = |val: &_| self.hasher.hash_one(val);
+        let hash = hasher(items);
+
+        let entry = self.offsets.entry(
+            hash,
+            |&offset| Self::read(&self.buffer, offset).unwrap() == items,
+            |&offset| hasher(Self::read(&self.buffer, offset).unwrap()),
+        );
+        (entry, &mut self.buffer)
```

Comment:
I assume returning the buffer here is for annoying borrow checker reasons?

Reply:
Yes indeed, as `Entry<'_, usize>` borrows from `&mut self`, and you wouldn’t be able to mutate `self.buffer` alongside it.